src/layers.c

```diff
@@ -37,55 +56,41 @@ layers_arrange(struct output *output)
 		return;
 	}
 
-	int nr_layers = sizeof(output->layers) / sizeof(output->layers[0]);
+	int nr_layers = sizeof(output->layer_tree) / sizeof(output->layer_tree[0]);
 	for (int i = 0; i < nr_layers; i++) {
-		struct lab_layer_surface *lab_layer_surface;
+		struct wlr_scene_tree *layer = output->layer_tree[i];
 
 		/*
-		 * First we go over the list of surfaces that have
-		 * exclusive_zone set (e.g. statusbars) because we have to
-		 * determine the usable area before processing regular layouts.
+		 * Process exclusive-zone clients before non-exclusive-zone
+		 * clients, so that the latter give way to the former regardless
+		 * of the order in which they were launched.
 		 */
-		wl_list_for_each(lab_layer_surface, &output->layers[i], link) {
-			struct wlr_scene_layer_surface_v1 *scene_layer_surface =
-				lab_layer_surface->scene_layer_surface;
-			if (scene_layer_surface->layer_surface->current.exclusive_zone) {
-				wlr_scene_layer_surface_v1_configure(
-					scene_layer_surface, &full_area, &usable_area);
-			}
-		}
+		arrange_one_layer(output, &full_area, &usable_area, layer, true);
+		arrange_one_layer(output, &full_area, &usable_area, layer, false);
 
-		/* Now we process regular layouts */
-		wl_list_for_each(lab_layer_surface, &output->layers[i], link) {
-			struct wlr_scene_layer_surface_v1 *scene_layer_surface =
-				lab_layer_surface->scene_layer_surface;
-			if (!scene_layer_surface->layer_surface->current.exclusive_zone) {
-				wlr_scene_layer_surface_v1_configure(
-					scene_layer_surface, &full_area, &usable_area);
-			}
-		}
-
-		wlr_scene_node_set_position(&output->layer_tree[i]->node,
-			scene_output->x, scene_output->y);
+		/* Set node position to account for output layout change */
+		wlr_scene_node_set_position(&layer->node, scene_output->x,
+			scene_output->y);
 	}
 
 	memcpy(&output->usable_area, &usable_area, sizeof(struct wlr_box));
```

Comment:
Although not changed in this PR, IMHO we should use `sizeof(output->usable_area)` here to make sure we don't have an out-of-bounds write whenever we decide to change from `struct wlr_box` to something else in the future.

Reply:
Done